src/PixelGraph.cpp

```diff
@@ -509,4 +509,337 @@ int PixelGraph::calculateValence(cv::Point p) {
+                if ((n->top != NULL && n->bottom != NULL) || (n->left != NULL && n->right != NULL)) {
+                    // straight line through cardinals
+                    fills[y][x] = FillType::FULL;
+                }
+                else if ((n->topLeft != NULL && n->bottomRight != NULL) || (n->topRight != NULL && n->bottomLeft != NULL)) {
+                    // true diagonal
+                    fills[y][x] = FillType::DIAGONAL;
+                }
+                else {
+                    // now we need to check for diagonals!
+                    Node* top = (y - 1.0 < 0) ? NULL : graph->at(y - 1.0)->at(x);
+                    Node* bottom = (y + 1.0 >= graph->size()) ? NULL : graph->at(y + 1.0)->at(x);
+
+                    bool nHasDiagonals = n->bottomLeft != NULL || n->bottomRight != NULL || n->topLeft != NULL || n->topRight != NULL;
+                    bool topDiagonals = top != NULL && (top->bottomLeft != NULL || top->bottomRight != NULL);
+                    bool bottomDiagonals = bottom != NULL && (bottom->topLeft != NULL || bottom->topRight != NULL);
+                    bool otherDiagonals = topDiagonals || bottomDiagonals;
+                    
+                    if (nHasDiagonals || otherDiagonals) {
+                        fills[y][x] = FillType::NO_CORNER;
+                    }
+                    else {
+                        fills[y][x] = FillType::FULL;
+                    }
+                }
+            }
+            else {
+                // valence 3+ all have very similar rules (2 has some special cases, so we handle it separately)
+                if ((n->top != NULL && n->bottom != NULL) || (n->left != NULL && n->right != NULL)) {
+                    // straight line through cardinals
+                    fills[y][x] = FillType::FULL;
+                }
+                else if ((n->topLeft != NULL && n->bottomRight != NULL) || (n->topRight != NULL && n->bottomLeft != NULL)) {
+                    // true diagonal
+                    fills[y][x] = FillType::NO_CORNER;
+                }
+                else {
+                    // now we need to check for diagonals that are NOT from this colour!
+                    Node* top = (y - 1.0 < 0 || n->top != NULL) ? NULL : graph->at(y - 1.0)->at(x);
+                    Node* bottom = (y + 1.0 >= graph->size() || n->bottom != NULL) ? NULL : graph->at(y + 1.0)->at(x);
+
+                    //bool nHasDiagonals = n->bottomLeft != NULL || n->bottomRight != NULL || n->topLeft != NULL || n->topRight != NULL;
+                    bool topDiagonals = top != NULL && (top->bottomLeft != NULL || top->bottomRight != NULL);
+                    bool bottomDiagonals = bottom != NULL && (bottom->topLeft != NULL || bottom->topRight != NULL);
+                    bool otherDiagonals = topDiagonals || bottomDiagonals;
+
+                    if (otherDiagonals) {
+                        fills[y][x] = FillType::NO_CORNER;

```

Comment:
Same here... aren't there still cases where it's missing more than one corner?

Reply:
There's a case for an elbow to be added here as well, so I'll do that
```
x x x
- x -
- - -
```